src/main/kotlin/nl/info/zac/authentication/OidcSessionService.kt

```diff
@@ -0,0 +1,119 @@
+import java.net.HttpURLConnection
+import java.net.URL
+import java.net.URLEncoder
+
+@Singleton
+@AllOpen
+@Suppress("LongParameterList")
+class OidcSessionService internal constructor(
+    private val userPrincipalFilter: UserPrincipalFilter,
+    private val httpSession: Instance<HttpSession>,
+    private val authServer: String,
+    private val authRealm: String,
+    private val clientId: String,
+    private val clientSecret: String,
+    private val connectionFactory: (String) -> HttpURLConnection
+) {
+    @Inject
+    constructor(
+        userPrincipalFilter: UserPrincipalFilter,
+        httpSession: Instance<HttpSession>,
+        @ConfigProperty(name = "AUTH_SERVER") authServer: String,
+        @ConfigProperty(name = "AUTH_REALM") authRealm: String,
+        @ConfigProperty(name = "AUTH_RESOURCE") clientId: String,
+        @ConfigProperty(name = "AUTH_SECRET") clientSecret: String
+    ) : this(
+        userPrincipalFilter,
+        httpSession,
+        authServer,
+        authRealm,
+        clientId,
+        clientSecret,
+        { url -> URL(url).openConnection() as HttpURLConnection }
+    )
+
+    companion object {
+        private const val HTTP_OK_MIN = HttpURLConnection.HTTP_OK
+        private const val HTTP_OK_MAX = 299
+    }
+
+    /**
+     * Refreshes the OIDC user session using the refresh token stored in the HttpSession.
+     * Updates the session with the new refresh token.
+     * Throws OidcSessionException on error.
+     */
+    @Suppress("TooGenericExceptionCaught", "ThrowsCount")
+    fun refreshUserSession() {
+        val session = httpSession.get()
+        val refreshToken = session.getAttribute("refresh_token") as? String
+        check(refreshToken != null) { "No refresh token found in session" }
+
+        val params =
+            "grant_type=refresh_token" +
+                "&refresh_token=${URLEncoder.encode(refreshToken, "UTF-8")}" +
+                "&client_id=${URLEncoder.encode(clientId, "UTF-8")}" +
+                "&client_secret=${URLEncoder.encode(clientSecret, "UTF-8")}"
+
+        val keycloakUrl = "$authServer/realms/$authRealm/protocol/openid-connect/token"
+
+        val connection = connectionFactory(keycloakUrl)
+        connection.requestMethod = "POST"
+        connection.doOutput = true
+        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
+        connection.outputStream.use { it.write(params.toByteArray()) }
+
+        if (connection.responseCode !in HTTP_OK_MIN..HTTP_OK_MAX) {
+            throw OidcSessionException("Failed to refresh user session: HTTP ${connection.responseCode}")
+        }
+
+        val response: String = try {
+            connection.inputStream.bufferedReader().readText()
+        } catch (e: Exception) {
+            val errorBody = connection.errorStream?.bufferedReader()?.readText()
+            val errorMessage = errorBody?.takeIf { it.isNotBlank() } ?: e.message
```

Comment:
Not sure this adds anything because the exception itself is already passed on to the `OidcSessionException` so the root exception message is already present in the `OidcSessionException`. We don't do this anywhere else I think?